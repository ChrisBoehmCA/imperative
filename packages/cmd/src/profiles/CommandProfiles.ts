/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeError } from "../../../error";
import { IProfile } from "../../../profiles";
import { ImperativeExpect } from "../../../expect";
import { isNullOrUndefined } from "util";
import { Imperative } from "../../../index";
/**
 * Profiles map created by the command profile loader and passed to the handler via parameters. Handlers can
 * retreive loaded profiles from the map via the profile type.
 * @export
 * @class CommandProfiles
 */
export class CommandProfiles {
    /**
     * The loaded profiles map - profiles are retrieved via the type key. More than one profile of a particular type
     * can be loaded by the command processor (depending on dependencies, etc.)
     * @private
     * @type {Map<string, IProfile[]>}
     * @memberof CommandProfiles
     */
    private mMap: Map<string, IProfile[]>;

    /**
     * Creates an instance of CommandProfiles.
     * @param {Map<string, IProfile[]>} map - The map of profiles
     * @memberof CommandProfiles
     */
    constructor(map: Map<string, IProfile[]>) {
        // Simple validation of input parameters
        const err: string = "Command Profiles Internal Error:";
        ImperativeExpect.toNotBeNullOrUndefined(map, `${err} No map was supplied.`);
        ImperativeExpect.toBeEqual(map instanceof Map, true, `${err} The "map" supplied is not an instance of a map.`);

        // Ensure the correctness of each map entry
        map.forEach((value, key) => {
            ImperativeExpect.toBeAnArray(value, `${err} The "profiles" supplied for type "${key}" is NOT an array.`);
            ImperativeExpect.toBeEqual((value.length > 0), true, `${err} No profiles supplied for type "${key}".`);
            for (const prof of value) {
                ImperativeExpect.keysToBeDefinedAndNonBlank(prof, ["name"],
                    `${err} A profile supplied in the map (type "${key}") does NOT have a name.`);
                ImperativeExpect.keysToBeDefinedAndNonBlank(prof, ["type"],
                    `${err} A profile supplied in the map (type "${key}") does NOT have a type.`);
                ImperativeExpect.toBeEqual(prof.type, key,
                    `${err} A profile supplied for type "${key}" indicates a type of "${prof.type}"`);
            }
        });
        this.mMap = map;
    }

    /**
     * Gets the first (or by name) profile in the map - automatically throws an exception (unless disabled)
     * @template T - The expected profile mapping to be returned
     * @param {string} type - The profile type
     * @param {string} [name=""] - The name of the profile to retrieve
     * @param {boolean} [failNotFound=true] - Automatically throws an imperative exception if not profiles are not
     * found - this is provided as convince for the handlers (will fail your command if not found) - This would
     * normally be the result of a command configuration problem.
     * @returns {T} - The first profile in the map (or the one located by name)
     * @memberof CommandProfiles
     */
    public get<T extends IProfile>(type: string, failNotFound = true, name = ""): IProfile {
        let profile;
        // If a profile is returned for the type, then we'll check if a profile of a specific name was requseted
        // if not, just return the first profile found (first loaded)
        if (this.map.get(type) != null) {
            if (name != null && name.trim().length > 0) {
                for (const prof of this.map.get(type)) {
                    if (prof.name === name) {
                        profile = prof;
                        break;
                    }
                }
            } else {
                profile = this.map.get(type)[0];
            }
        } else if (failNotFound) {
            this.fail(type);
        }
        return profile;
    }

    /**
     * Gets all profiles for the type specified,
     * @template T - extends IProfile
     * @param {string} type - The profile type to retrieve loaded profiles
     * @returns {T[]} - The list of profile types
     * @param {boolean} [failNotFound=true] - Automatically throws an imperative exception if not profiles are not
     * found - this is provided as convince for the handlers (will fail your command if not found) - This would
     * normally be the result of a command configuration problem.
     * @memberof CommandProfiles
     */
    public getAll<T extends IProfile>(type: string, failNotFound = true): IProfile[] {
        const profiles = this.map.get(type);
        if (profiles == null && failNotFound) {
            this.fail(type);
        }
        return profiles;
    }

    /**
     * Internal accessor for the map
     * @readonly
     * @private
     * @type {Map<string, IProfile[]>} - The profile Map
     * @memberof CommandProfiles
     */
    private get map(): Map<string, IProfile[]> {
        return this.mMap;
    }

    /**
     * Throw an error failing the get(requested by the caller if no profiles are available)
     * @private
     * @param {string} type - the profile type to get from the map
     * @memberof CommandProfiles
     */
    private fail(type: string) {
        throw new ImperativeError({
            msg: `Internal Error: No profiles of type "${type}" were loaded for this command.`,
            additionalDetails: `This error can occur for one of two reasons:\n` +
                ` - The "profile" property on the command definition document ` +
                `does NOT specify the requested profile type\n` +
                ` - The profile type is marked "optional", no profiles of type "${type}" have been created, ` +
                `and the command handler requested a profile of type "${type}" with "failNotFound=true"`
        });
    }
}
