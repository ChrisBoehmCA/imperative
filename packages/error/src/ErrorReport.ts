/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IErrorReportParms } from "./../../error";
import { isNullOrUndefined } from "util";
import { ImperativeError } from "./ImperativeError";
import { IErrorSavedSettings } from "./doc/IErrorSavedSettings";

/**
 * Class to make use of node-report as an optional dependency.  This is optional
 * because it is a native module with native dependencies.
 * @export
 * @class ErrorReport
 */
export class ErrorReport {

    /**
     * Message for uninstalled node-report
     * @static
     * @memberof ErrorReport
     */
    public static readonly NOT_INSTALLED_MESSAGE = "Node-report is not installed";

    /**
     * Return a report to the called to be displayed or logged in the format of the
     * caller's choosing.
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @returns {string} - returns string report
     * @memberof ErrorReport
     */
    public static obtain(parms?: IErrorReportParms): string {
        const settings = ErrorReport.setEnvironment(parms);
        let report: string;
        try {
            const nodeReport = require("node-report");
            report = nodeReport.getReport();
        } catch (error) {
            report = ErrorReport.NOT_INSTALLED_MESSAGE;
        } finally {
            ErrorReport.restoreEnvironment(settings);
        }
        return report;
    }

    /**
     * Trigger a report to be generated as a text document.
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @memberof ErrorReport
     */
    public static trigger(parms?: IErrorReportParms) {
        const settings = ErrorReport.setEnvironment(parms);
        try {
            const nodeReport = require("node-report");
            nodeReport.triggerReport();
        } catch (error) {
            // return ErrorReport.NOT_INSTALLED_MESSAGE;
        } finally {
            ErrorReport.restoreEnvironment(settings);
        }
    }

    /**
     * Set environment variables for node report prior to using it.
     * @private
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @memberof ErrorReport
     */
    private static setEnvironment(parms?: IErrorReportParms) {
        const settings: IErrorSavedSettings = {};
        settings.NODEREPORT_VERBOSE = "no";
        if (!isNullOrUndefined(parms)) {
            if (!isNullOrUndefined(parms.verbose)) {
                process.env.NODEREPORT_VERBOSE = parms.verbose;
            }
        }
        return settings;
    }

    /**
     * Restore environmental settings
     * @private
     * @static
     * @param {IErrorSavedSettings} settings - object to contain restoration values for env variable
     * @memberof ErrorReport
     */
    private static restoreEnvironment(settings: IErrorSavedSettings) {
        process.env.NODEREPORT_VERBOSE = settings.NODEREPORT_VERBOSE;
    }
}
