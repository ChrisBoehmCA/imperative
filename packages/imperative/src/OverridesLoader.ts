/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*/

import {IImperativeOverrides} from "./doc/IImperativeOverrides";
import { CredentialManagerFactory, DefaultCredentialManager } from "../../security";
import {IImperativeConfig} from "./doc/IImperativeConfig";
import { isAbsolute, resolve } from "path";

/**
 * Imperative-internal class to load overrides
 * You should not need to call this from your CLI.
 */
export class OverridesLoader {
  /**
   * Apply overrides to all applicable facilities and use our defaults where
   * an override is not provided.
   *
   * @param {IImperativeConfig} config - the current {@link Imperative#loadedConfig}
   */
  public static async load(
    config: IImperativeConfig
  ): Promise<void> {
    // Initialize the Credential Manager
    await this.loadCredentialManager(config);
  }

  /**
   * Initialize the Credential Manager using the supplied override when provided.
   *
   * @param {IImperativeConfig} config - the current {@link Imperative#loadedConfig}
   */
  private static async loadCredentialManager(
    config: IImperativeConfig
  ): Promise<void> {
    const overrides: IImperativeOverrides = config.overrides;

    // If the credential manager wasn't set, then we use the DefaultCredentialManager
    if (overrides.CredentialManager == null) {
      overrides.CredentialManager = DefaultCredentialManager;
    }
    // If the credential manager is type string and not absolute, we will convert it to an absolute path
    // relative to the process entry file location.
    else if (typeof overrides.CredentialManager === "string" && !isAbsolute(overrides.CredentialManager)) {
      overrides.CredentialManager = resolve(process.mainModule.filename, "../", overrides.CredentialManager);
    }

    await CredentialManagerFactory.initialize(overrides.CredentialManager, config.name);
  }
}
