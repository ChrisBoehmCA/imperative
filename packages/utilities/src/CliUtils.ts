/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { isNullOrUndefined } from "util";
import { ImperativeError } from "../../error";
import { Constants } from "../../constants";
import { Arguments } from "yargs";
import { TextUtils } from "./TextUtils";
import { IOptionFormat } from "./doc/IOptionFormat";
import { Logger } from "../../logger";

/**
 * Cli Utils contains a set of static methods/helpers that are CLI related (forming options, censoring args, etc.)
 * @export
 * @class CliUtils
 */
export class CliUtils {
    /**
     * Used as the place holder when censoring arguments in messages/command output
     * @static
     * @memberof CliUtils
     */
    public static readonly CENSOR_RESPONSE = "****";

    /**
     * A list of cli options/keywords that should normally be sensored
     * @static
     * @memberof CliUtils
     */
    public static CENSORED_OPTIONS = ["auth", "p", "pass", "password", "passphrase", "credentials",
        "authentication", "basic-auth", "basicAuth"];

    /**
     * Get the 'dash form' of an option as it would appear in a user's command,
     * appending the proper number of dashes depending on the length of the option name
     * @param {string} optionName - e.g. my-option
     * @returns {string} - e.g. --my-option
     */
    public static getDashFormOfOption(optionName: string): string {
        if (!isNullOrUndefined(optionName) && optionName.length >= 1) {
            const dashes = optionName.length > 1 ? Constants.OPT_LONG_DASH : Constants.OPT_SHORT_DASH;
            return dashes + optionName;
        }
        else {
            throw new ImperativeError({
                msg: "A null or blank option was supplied. Please correct the option definition."
            });
        }
    }

    /**
     * Copy and censor any sensitive CLI arguments before logging/printing
     * @param {string[]} args - The args list to censor
     * @returns {string[]}
     */
    public static censorCLIArgs(args: string[]): string[] {
        const newArgs: string[] = JSON.parse(JSON.stringify(args));
        const censoredValues = CliUtils.CENSORED_OPTIONS.map(CliUtils.getDashFormOfOption);
        for (const value of censoredValues) {
            if (args.indexOf(value) >= 0) {
                const valueIndex = args.indexOf(value);
                if (valueIndex < args.length - 1) {
                    newArgs[valueIndex + 1] = CliUtils.CENSOR_RESPONSE; // censor the argument after the option name
                }
            }
        }
        return newArgs;
    }

    /**
     * Copy and censor a yargs argument object before logging
     * @param {yargs.Arguments} args the args to censor
     * @returns {yargs.Arguments}  a censored copy of the arguments
     */
    public static censorYargsArguments(args: Arguments): Arguments {
        const newArgs: Arguments = JSON.parse(JSON.stringify(args));

        for (const optionName of Object.keys(newArgs)) {
            if (CliUtils.CENSORED_OPTIONS.indexOf(optionName) >= 0) {
                const valueToCensor = newArgs[optionName];
                newArgs[optionName] = CliUtils.CENSOR_RESPONSE;
                for (const checkAliasKey of Object.keys(newArgs)) {
                    if (newArgs[checkAliasKey] === valueToCensor) {
                        newArgs[checkAliasKey] = CliUtils.CENSOR_RESPONSE;
                    }
                }
            }
        }
        return newArgs;
    }

    /**
     * Get the value of an environment variable associated with the specified option name.
     * The environment variable name will be formed by concatenating an environment name prefix,
     * the subCmdName, and the cmdOption using underscores as delimeters.
     *
     * The subCmdName and cmdOption name can be specified in camelCase or in kabab-style.
     * Regardless of the style, both will be converted to upper case
     * We replace dashes in Kabab-style values with underscores. We replace each uppercase
     * character in a camelCase value with underscore and that character.
     *
     * The envPrefix will be used exactly as specified.
     *
     * Example: The values myEnv-Prefix, some-command-name someOptionName would retrieve
     * the value of an environment variable named
     *      myEnv-Prefix_SOME_COMMAND_NAME_SOME_OPTION_NAME
     *
     * @param {string} envPrefix - The prefix for environment variables for this CLI.
     *      The value obtained by ImperativeConfig.getEnvVariablePrefix() will use the
     *      envVariablePrefix from the Imperative config object, and will use the
     *      rootCommandName as a fallback value.
     *
     * @param {string} subCmdName - The name of the 1st-level sub-command that is being run.
     *
     * @param {string} cmdOption - The name of the option in either camelCase or kabab-style.
     *
     * @returns {string | null} - The value of the environment variable which corresponds
     *      to the supplied option for the supplied command. If no such environment variable
     *      exists we return null.
     *
     * @memberof CliUtils
     */
    public static getEnvValForOption(
        envPrefix: string, subCmdName: string, cmdOption: string
    ): string | null
    {
        const cmdNmChoices: IOptionFormat = CliUtils.getOptionFormat(subCmdName);
        const optChoices: IOptionFormat = CliUtils.getOptionFormat(cmdOption);

        // Form envPrefix, subCmdName, and option into an environment variable
        const envDelim = "_";
        let envVarName = cmdNmChoices.kebabCase + envDelim + optChoices.kebabCase;
        envVarName = envPrefix + envDelim + envVarName.toUpperCase().replace(/-/g, envDelim);

        // Get the value of the environment variable
        if (process.env.hasOwnProperty(envVarName)) {
            return process.env[envVarName];
        }

        // no corresponding environment variable exists
        return null;
    }

    /**
     * Constructs the yargs style positional argument string.
     * @static
     * @param {boolean} positionalRequired - Indicates that this positional is required
     * @param {string} positionalName - The name of the positional
     * @returns {string} - The yargs style positional argument string (e.g. <name>);
     * @memberof CliUtils
     */
    public static getPositionalSyntaxString(positionalRequired: boolean, positionalName: string): string {
        const leftChar = positionalRequired ? "<" : "[";
        const rightChar = positionalRequired ? ">" : "]";
        return leftChar + positionalName + rightChar;
    }

    /**
     * Format the help header - normally used in help generation etc.
     * @static
     * @param {string} header
     * @param {string} [indent=" "]
     * @param {string} color
     * @returns {string}
     * @memberof CliUtils
     */
    public static formatHelpHeader(header: string, indent: string = " ", color: string): string {
        if (isNullOrUndefined(header) || header.trim().length === 0) {
            throw new ImperativeError({
                msg: "Null or empty header provided; could not be formatted."
            });
        }
        const numDashes = header.length + 1;
        const headerText = TextUtils.formatMessage("{{indent}}{{headerText}}\n{{indent}}{{dashes}}",
            { headerText: header.toUpperCase(), dashes: Array(numDashes).join("-"), indent });
        return TextUtils.chalk[color](headerText);
    }

    /**
     * Takes a key and converts it to both camelCase and kebab-case.
     *
     * @param key The key to transform
     *
     * @returns An object that contains the new format.
     *
     * @example <caption>Conversion of keys</caption>
     *
     * CliUtils.getOptionFormat("helloWorld");
     *
     * // returns
     * const return1 = {
     *     key: "helloWorld",
     *     camelCase: "helloWorld",
     *     kebabCase: "hello-world"
     * }
     *
     * /////////////////////////////////////////////////////
     *
     * CliUtils.getOptionFormat("hello-world");
     *
     * // returns
     * const return2 = {
     *     key: "hello-world",
     *     camelCase: "helloWorld",
     *     kebabCase: "hello-world"
     * }
     *
     * /////////////////////////////////////////////////////
     *
     * CliUtils.getOptionFormat("hello--------world");
     *
     * // returns
     * const return3 = {
     *     key: "hello--------world",
     *     camelCase: "helloWorld",
     *     kebabCase: "hello-world"
     * }
     *
     * /////////////////////////////////////////////////////
     *
     * CliUtils.getOptionFormat("hello-World-");
     *
     * // returns
     * const return4 = {
     *     key: "hello-World-",
     *     camelCase: "helloWorld",
     *     kebabCase: "hello-world"
     * }
     */
    public static getOptionFormat(key: string): IOptionFormat {
        return {
            camelCase: key.replace(/(-+\w?)/g, (match, p1) => {
                /*
                 * Regular expression checks for 1 or more "-" characters followed by 0 or 1 word character
                 * The last character in each match is converted to upper case and returned only if it
                 * isn't equal to "-"
                 *
                 * Examples: (input -> output)
                 *
                 * - helloWorld         -> helloWorld
                 * - hello-world        -> helloWorld
                 * - hello--------world -> helloWorld
                 * - hello-World-       -> helloWorld
                 */
                const returnChar = p1.substr(-1).toUpperCase();
                return  returnChar !== "-" ? returnChar : "";
            }),
            kebabCase: key.replace(/(-*[A-Z]|-{2,}|-$)/g, (match, p1, offset, inputString) => {
                /*
                 * Regular expression matches the following:
                 *
                 * 1. Any string segment containing 0 or more "-" characters followed by any uppercase letter.
                 * 2. Any string segment containing 2 or more consecutive "-" characters
                 * 3. Any string segment where the last character is "-"
                 *
                 * Matches for 1.
                 *
                 * - "A"           -> If condition 1.2
                 * - "-B"          -> If condition 2.2
                 * - "------C"     -> If condition 2.2
                 *
                 * Matches for 2.
                 *
                 * - "--"          -> If condition 2.1.1
                 * - "-------"     -> If condition 2.1.1 or 2.1.2
                 *
                 * 2.1.1 will be entered if the match is the last sequence of the string
                 * 2.1.2 will be entered if the match is not the last sequence of the string
                 *
                 * Matches for 3.
                 * - "-<end_of_string>" -> If condition 1.1
                 *
                 * Examples: (input -> output)
                 *
                 * - helloWorld         -> hello-world
                 * - hello-world        -> hello-world
                 * - hello--------world -> hello-world
                 * - hello-World-       -> hello-world
                 */

                if (p1.length === 1) {                                          // 1
                    if (p1 === "-") {                                           // 1.1
                        // Strip trailing -
                        return "";
                    } else {                                                    // 1.2
                        // Change "letter" to "-letter"
                        return "-" + p1.toLowerCase();
                    }
                } else {                                                        // 2
                    const returnChar = p1.substr(-1); // Get the last character of the sequence

                    if (returnChar === "-") {                                   // 2.1
                        if (offset + p1.length === inputString.length) {        // 2.1.1
                            // Strip a trailing -------- sequence
                            return "";
                        } else {                                                // 2.1.2
                            // Change a sequence of -------- to a -
                            return "-";
                        }
                    } else {                                                    // 2.2
                        // Change a sequence of "-------letter" to "-letter"
                        return "-" + returnChar.toLowerCase();
                    }
                }
            }),
            key
        };
    }
}
