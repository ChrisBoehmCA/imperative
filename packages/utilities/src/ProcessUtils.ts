/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeError } from "../../error";
import { isNullOrUndefined } from "util";

/**
 * A collection of utilities related to the running process.
 * @export
 * @class ProcessUtils
 */
export class ProcessUtils {
    // __________________________________________________________________________
    /**
     * Process utility to wrap callback process routines into promises
     * Turn nextTick into a promise to prevent nesting
     * @static
     * @param {() => void} callback - called before promise is resolved
     * @param {...any[]} args - arguments passed to the callback
     * @returns {Promise<void>} - fullfilled whenever callback is invoked
     * @memberof ProcessUtils
     */
    public static nextTick(callback: (...args: any[]) => void, ...args: any[]): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            process.nextTick(() => {
                callback(...args);
                resolve();
            });
        });
    }

    // __________________________________________________________________________
    /**
     * Is a Graphical User Interface avaiable in the environment in which
     * the current command is running?
     *
     * @returns {boolean} - True if GUI. False when no GUI.
     */
    public static isGuiAvailable(): boolean {
        /* If any of the SSH environment variables are defined,
         * then we are in an ssh session --> no GUI.
         */
        if (typeof process.env.SSH_CONNECTION !== "undefined" ||
            typeof process.env.SSH_CLIENT !== "undefined" ||
            typeof process.env.SSH_TTY !== "undefined")
        {
            return false;
        }

        /* On linux (and MAC?) the DISPLAY environment variable
         * indicates that we are in an X-Windows environment.
         */
        if ( process.platform !== "win32") {
            if (typeof process.env.DISPLAY === "undefined" ||
                process.env.DISPLAY === "" ) {
                return false;
            } else {
                return true;
            }
        }

        // otherwise we assume we have a GUI
        return true;
    }
}
