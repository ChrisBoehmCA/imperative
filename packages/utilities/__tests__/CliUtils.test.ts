/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { CliUtils } from "../src/CliUtils";
import { CommandProfiles, ICommandOptionDefinition } from "../../cmd";
import { IProfile } from "../../profiles";
import { ImperativeError } from "../../error";

describe("CliUtils", () => {
    describe("buildBaseArgs", () => {
        it("should preserve the _ and $0 properties", () => {
            const args = CliUtils.buildBaseArgs({_: ["cmd1", "cmd2"], $0: "test exe"});
            expect(args).toMatchSnapshot();
        });

        it("should remove properties that are set to undefined", () => {
            const args = CliUtils.buildBaseArgs({_: ["cmd1", "cmd2"], $0: "test exe", test: undefined});
            expect(args).toMatchSnapshot();
        });

        it("should preserve already set properties (that are not undefined)", () => {
            const args = CliUtils.buildBaseArgs({_: ["cmd1", "cmd2"], $0: "test exe", test: true});
            expect(args).toMatchSnapshot();
        });
    });
    describe("getOptValueFromProfiles", () => {

        const FAKE_OPTS: ICommandOptionDefinition[] = [{
            name: "fake-string-opt",
            description: "a fake opt",
            type: "string"
        }, {
            name: "nohyphen",
            description: "a fake opt",
            type: "string"
        }, {
            name: "couldBeEither",
            description: "a fake opt",
            type: "string"
        },
        {
            name: "with-alias",
            description: "a fake opt",
            type: "string",
            aliases: ["w"]
        }];

        it("should throw an imperative error if a required profile is not present", () => {
            let error;
            try {
                const args = CliUtils.getOptValueFromProfiles(
                    new CommandProfiles(new Map<string, IProfile[]>()),
                    { required: ["banana"] },
                    FAKE_OPTS);
            } catch (e) {
                error = e;
            }
            expect(error).toBeDefined();
            expect(error instanceof ImperativeError).toBe(true);
            expect(error.message).toMatchSnapshot();
        });

        it("should return nothing if a profile was optional and not loaded", () => {
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(new Map<string, IProfile[]>()),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(Object.keys(args).length).toBe(0);
        });

        it("should return args (from definitions with no hyphen in name) extracted from loaded profile", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{ type: "banana", name: "fakebanana", nohyphen: "specified in profile" }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });

        it("should return args (with both cases) extracted from loaded profile, preferring the camel case", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{
                "type": "banana",
                "name": "fakebanana",
                "couldBeEither": "should be me",
                "could-be-either": "should not be me"
            }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });

        it("should return args (with both cases) extracted from loaded profile, preferring the kebab case", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{
                "type": "banana",
                "name": "fakebanana",
                "fakeStringOpt": "should not be me",
                "fake-string-opt": "should be me"
            }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });

        it("should return args with both cases, if the option is camel and the profile is kebab", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{
                "type": "banana",
                "name": "fakebanana",
                "could-be-either": "should be me"
            }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });

        it("should return args with both cases, if the option is kebab and the profile is camel", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{
                type: "banana",
                name: "fakebanana",
                fakeStringOpt: "should be me"
            }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });

        it("should return args with aliases if extracted option from a profile", () => {
            const map = new Map<string, IProfile[]>();
            map.set("banana", [{
                type: "banana",
                name: "fakebanana",
                withAlias: "should have 'w' on args object too"
            }]);
            const args = CliUtils.getOptValueFromProfiles(
                new CommandProfiles(map),
                { optional: ["banana"] },
                FAKE_OPTS);
            expect(args).toMatchSnapshot();
        });
    });

    it("should be able to produce the --dash-form of any options", () => {
        // correct uses
        expect(CliUtils.getDashFormOfOption("some-option")).toMatchSnapshot();
        expect(CliUtils.getDashFormOfOption("s")).toMatchSnapshot();

        // edge cases (nothing breaks, no error, just tons of dashes)
        expect(CliUtils.getDashFormOfOption("-some-option")).toMatchSnapshot();
        expect(CliUtils.getDashFormOfOption("--some-option")).toMatchSnapshot();
        expect(CliUtils.getDashFormOfOption("-s")).toMatchSnapshot();
        expect(CliUtils.getDashFormOfOption("--s")).toMatchSnapshot();

        // empty cases
        expect(() => {
            CliUtils.getDashFormOfOption("");
        }).toThrowErrorMatchingSnapshot();

    });

    it("should return the proper option format", () => {
        expect(CliUtils.getOptionFormat("hello-world")).toEqual({
            key: "hello-world",
            camelCase: "helloWorld",
            kebabCase: "hello-world"
        });

        expect(CliUtils.getOptionFormat("helloWorld")).toEqual({
            key: "helloWorld",
            camelCase: "helloWorld",
            kebabCase: "hello-world"
        });

        expect(CliUtils.getOptionFormat("hello--------world")).toEqual({
            key: "hello--------world",
            camelCase: "helloWorld",
            kebabCase: "hello-world"
        });

        expect(CliUtils.getOptionFormat("hello-World-")).toEqual({
            key: "hello-World-",
            camelCase: "helloWorld",
            kebabCase: "hello-world"
        });

        expect(CliUtils.getOptionFormat("hello-World-----------")).toEqual({
            key: "hello-World-----------",
            camelCase: "helloWorld",
            kebabCase: "hello-world"
        });
    });

    it("should return environment value for kabab-style option", () => {
        const expectedEnvVarValue = "The value for kabab-style option";
        process.env.MYENVPREFIX_OPT_MY_OPTION = expectedEnvVarValue;

        const recvEnvValue = CliUtils.getEnvValForOption("MYENVPREFIX",
            "my-option"
        );
        expect(recvEnvValue).toEqual(expectedEnvVarValue);
    });

    it("should return environment value for camelCase option", () => {
        const expectedEnvVarValue = "The value for camelCase option";
        process.env.MYENVPREFIX_OPT_MY_OPTION = expectedEnvVarValue;

        const recvEnvValue = CliUtils.getEnvValForOption("MYENVPREFIX",
            "myOption"
        );
        expect(recvEnvValue).toEqual(expectedEnvVarValue);
    });

    it("should not alter the environment prefix", () => {
        const expectedEnvVarValue = "The value for camelCase-kabab prefix";
        process.env["camelCase-kabab-Prefix_OPT_MY_OPTION"] = expectedEnvVarValue;

        const recvEnvValue = CliUtils.getEnvValForOption("camelCase-kabab-Prefix",
            "my-option"
        );
        expect(recvEnvValue).toEqual(expectedEnvVarValue);
    });

    it("should return NULL when environment variable does not exist", () => {
        const recvEnvValue = CliUtils.getEnvValForOption("MYENVPREFIX",
            "not-set-in-env"
        );
        expect(recvEnvValue).toEqual(null);
    });
});
