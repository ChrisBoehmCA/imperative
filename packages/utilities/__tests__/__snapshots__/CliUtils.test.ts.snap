// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CliUtils buildBaseArgs should preserve already set properties (that are not undefined) 1`] = `
Object {
  "$0": "test exe",
  "_": Array [
    "cmd1",
    "cmd2",
  ],
  "test": true,
}
`;

exports[`CliUtils buildBaseArgs should preserve the _ and $0 properties 1`] = `
Object {
  "$0": "test exe",
  "_": Array [
    "cmd1",
    "cmd2",
  ],
}
`;

exports[`CliUtils buildBaseArgs should remove properties that are set to undefined 1`] = `
Object {
  "$0": "test exe",
  "_": Array [
    "cmd1",
    "cmd2",
  ],
}
`;

exports[`CliUtils getOptValueFromProfiles should return args (from definitions with no hyphen in name) extracted from loaded profile 1`] = `
Object {
  "nohyphen": "specified in profile",
}
`;

exports[`CliUtils getOptValueFromProfiles should return args (with both cases) extracted from loaded profile, preferring the camel case 1`] = `
Object {
  "could-be-either": "should be me",
  "couldBeEither": "should be me",
}
`;

exports[`CliUtils getOptValueFromProfiles should return args (with both cases) extracted from loaded profile, preferring the kebab case 1`] = `
Object {
  "fake-string-opt": "should be me",
  "fakeStringOpt": "should be me",
}
`;

exports[`CliUtils getOptValueFromProfiles should return args with aliases if extracted option from a profile 1`] = `
Object {
  "w": "should have 'w' on args object too",
  "with-alias": "should have 'w' on args object too",
  "withAlias": "should have 'w' on args object too",
}
`;

exports[`CliUtils getOptValueFromProfiles should return args with both cases, if the option is camel and the profile is kebab 1`] = `
Object {
  "could-be-either": "should be me",
  "couldBeEither": "should be me",
}
`;

exports[`CliUtils getOptValueFromProfiles should return args with both cases, if the option is kebab and the profile is camel 1`] = `
Object {
  "fake-string-opt": "should be me",
  "fakeStringOpt": "should be me",
}
`;

exports[`CliUtils getOptValueFromProfiles should throw an imperative error if a required profile is not present 1`] = `"Profile of type \\"banana\\" does not exist within the loaded profiles for the command and it is marked as required."`;

exports[`CliUtils should be able to produce the --dash-form of any options 1`] = `"--some-option"`;

exports[`CliUtils should be able to produce the --dash-form of any options 2`] = `"-s"`;

exports[`CliUtils should be able to produce the --dash-form of any options 3`] = `"---some-option"`;

exports[`CliUtils should be able to produce the --dash-form of any options 4`] = `"----some-option"`;

exports[`CliUtils should be able to produce the --dash-form of any options 5`] = `"---s"`;

exports[`CliUtils should be able to produce the --dash-form of any options 6`] = `"----s"`;

exports[`CliUtils should be able to produce the --dash-form of any options 7`] = `"A null or blank option was supplied. Please correct the option definition."`;
